test/unit/test_ordered_forest.cpp

```diff
@@ -0,0 +1,497 @@
+
+        ASSERT_TRUE(forest_invariant(f));
+
+        ASSERT_EQ(6u, f.size());
+        EXPECT_EQ(12u, alloc.n_alloc()); // six nodes, six items.
+
+        auto i = f.begin();
+        ASSERT_TRUE(i);
+        EXPECT_EQ(1, *i);
+        EXPECT_FALSE(i.child());
+
+        i = i.next();
+        ASSERT_TRUE(i);
+        auto j = i.child();
+        ASSERT_TRUE(j);
+        EXPECT_EQ(4, *j);
+        j = j.next();
+        ASSERT_TRUE(j);
+        EXPECT_EQ(5, *j);
+        j = j.next();
+        ASSERT_TRUE(j);
+        EXPECT_EQ(6, *j);
+        EXPECT_FALSE(j.next());
+
+        i = i.next();
+        ASSERT_TRUE(i);
+        EXPECT_EQ(3, *i);
+        EXPECT_FALSE(i.child());
+        EXPECT_FALSE(i.next());
+    }
+
+    EXPECT_EQ(alloc.n_dealloc(), alloc.n_alloc());
+}
+
+TEST(ordered_forest, initializer_list) {
+    ordered_forest<int> f = {1, {2, {4, 5, 6}}, 3};
+    EXPECT_EQ(6u, f.size());
+
+    ASSERT_TRUE(forest_invariant(f));
+
+    auto i = f.begin();
+    ASSERT_TRUE(i);
+    EXPECT_EQ(1, *i);
+    EXPECT_FALSE(i.child());
+
+    i = i.next();
+    ASSERT_TRUE(i);
+    auto j = i.child();
+    ASSERT_TRUE(j);
+    EXPECT_EQ(4, *j);
+    j = j.next();
+    ASSERT_TRUE(j);
+    EXPECT_EQ(5, *j);
+    j = j.next();
+    ASSERT_TRUE(j);
+    EXPECT_EQ(6, *j);
+    EXPECT_FALSE(j.next());
+
+    i = i.next();
+    ASSERT_TRUE(i);
+    EXPECT_EQ(3, *i);
+    EXPECT_FALSE(i.child());
+    EXPECT_FALSE(i.next());
+}
+
+TEST(ordered_forest, equality) {
+    using of = ordered_forest<int>;
+
+    EXPECT_EQ(of{}, of{});
+    EXPECT_NE(of{1},  of{});
+    EXPECT_NE(of{},  of{1});
+
+    EXPECT_EQ((of{1, 2, 3}), (of{1, 2, 3}));
+    EXPECT_NE((of{1, 2, 3}),  (of{1, {2, {3}}}));
+    EXPECT_NE((of{{1, {2, 3}}}),  (of{1, 2, 3}));
+
+    struct always_eq {
+        int n_;
+        always_eq(int n): n_(n) {}
+        bool operator==(const always_eq&) const { return true; }
+        bool operator!=(const always_eq&) const { return false; }
+    };
+
+    ordered_forest<always_eq> f1 = {{1, {2, 3}}, {4, {5, {6, {7}}, 8}}, 9};
+    ordered_forest<always_eq> f2 = {{3, {1, 0}}, {2, {8, {6, {4}}, 7}}, 5};
+    EXPECT_EQ(f2, f1);
+
+    ordered_forest<always_eq> f3 = {{3, {{1, {0}}}}, {2, {8, {6, {4}}, 7}}, 5};
+    EXPECT_NE(f1, f3);
+}
+
+TEST(ordered_forest, iteration) {
+    using ivector = std::vector<int>;
+
+    ordered_forest<int> f = {{1, {2, 3}}, {4, {5, {6, {7}}, 8}}, 9};
+
+    ivector pre{f.preorder_begin(), f.preorder_end()};
+    EXPECT_EQ((ivector{1, 2, 3, 4, 5, 6, 7, 8, 9}), pre);
+
+    ivector pre_default{f.begin(), f.end()};
+    EXPECT_EQ((ivector{1, 2, 3, 4, 5, 6, 7, 8, 9}), pre_default);
+
+    ivector pre_cdefault{f.cbegin(), f.cend()};
+    EXPECT_EQ((ivector{1, 2, 3, 4, 5, 6, 7, 8, 9}), pre_cdefault);
+
+    ivector root{f.root_begin(), f.root_end()};
+    EXPECT_EQ((ivector{1, 4, 9}), root);
+
+    ivector post{f.postorder_begin(), f.postorder_end()};
+    EXPECT_EQ((ivector{2, 3, 1, 5, 7, 6, 8, 4, 9}), post);
+
```

Comment:
Missing tests for const root iterators and const postorder iterators, which allowed that const postorder iterator begin bug to slip by...